### DISCUSSION TITLE
Question about Hazelcast usage on a global rate limit

### DISCUSSION THREAD

**Discussion by fombico** [vladimir-bukhtoyarov/bucket4j#173] [Q&A]
Hi there,

I am using Hazelcast `4.0.3` with `bucket4j-hazelcast:6.2.0` to implement global rate limits per endpoint.

After running load tests, I've noticed two things with my implementation:

1. For a single global rate limit, the data is stored on the same Hazelcast partition. [However, all the operations of a partition are owned by a single Hazelcast operation thread.](https://docs.hazelcast.com/imdg/4.2/performance/threading-model.html#operation-threading) All the requests would wait on this one thread at some point, leading to performance issues under high load.

2. Without specifying a partitioning strategy, Hazelcast seems to use `DefaultPartitioningStrategy` by default. When the hashing algorithm in `com.hazelcast.internal.serialization.impl.AbstractSerializationService.calculatePartitionHash()` uses the object to determine a `partitionHash`, the result will be `0` as the object does not implement `PartitionAware`. The result is that even if I have rate limits on multiple endpoints, their objects get mapped to the same partition hash of `0`, stored in the same partition, and thus would all use the same operation thread.

I'm wondering if my understanding is correct, or perhaps I've misconfigured something? Ideally, I would not want all the requests to block on the same thread.

**Comment by vladimir-bukhtoyarov** [comment]
Hello @fombico 

About the first question: yes huge request rate on the same key is a problem. There [is a solution](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/7.0/bucket4j-core/src/main/java/io/github/bucket4j/distributed/proxy/optimization/batch/BatchingOptimization.java) that is implemented in the scope of the upcoming 7.0.0 release. It is example how to apply this optimization https://github.com/vladimir-bukhtoyarov/bucket4j/blob/7.0/bucket4j-examples/src/test/java/example/distributed/optimizers/batching/HazelcastWithBatchingPerformanceExample.java#L93
You can use **7.0.0-beta-2020-04-18** version to get access to this functionality.  I hope that benchmarks with BatchingOptimization will show better results.

About PartitionAware: if I understand correctly it is used for the collocation of logically coupled data on the same node, because state of each bucket is independent this functionality should be neutral for Bucket4j, but of course, it is better to consult with Hazelcast developers, maybe I miss something. And, you can do a quick test, by implementing PartitionAware for your keys and compare behavior with previous.

**Reply by fombico** [reply]
Thanks, I will take a look at 7.0.0-beta-2020-04-18.

Regarding PartitionAware, I was looking into this to avoid grouping them into the same partition. The data is [partitioned](https://docs.hazelcast.com/imdg/4.2/overview/data-partitioning.html#how-the-data-is-partitioned) using [this logic](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/AbstractSerializationService.java#L471) but the [DefaultParitioningStrategy](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/partition/strategy/DefaultPartitioningStrategy.java) will return null if PartitionAware is not implemented. Hence everything going to the same partition.

Anyway, I think I can mitigate the second issue by defining a custom `PartitioningStrategy` so that my routes result in different partitions.
